nautobot_circuit_maintenance/views.py

```diff
@@ -1,20 +1,197 @@
+        # template.
+        self.extra_content = {
+            "upcoming_maintenances": upcoming_days_maintenances,
+            "circuit_maint_metric_data": metric_values,
+            "n_days": n_days,
+        }
+
+    def extra_context(self):
+        """Extra content method on."""
+        # add global aggregations to extra context.
+
+        return self.extra_content
+
+    def get_maintenances_next_n_days(self, n_days: int):
+        """Gets maintenances in the next n number of days.
+
+        Args:
+            n_days (int): Number of days up coming
+
+        Returns:
+            Set: Set of maintenances that are up coming
+        """
+        today = datetime.date.today()
+        end_date = today + datetime.timedelta(days=n_days)
+        maintenances = self.queryset
+        return_list = []
+        for maintenance in maintenances:
+            if today <= maintenance.start_time.date() <= end_date:
+                return_list.append(maintenance)
+
+        return return_list
+
+    def get_maintenance_past_n_days(self, n_days: int):
+        """Gets maintenances in the past n number of days.
+
+        Args:
+            n_days (int): Should be a negative number for the number of days.
+        """
+        today = datetime.date.today()
+        end_date = today + datetime.timedelta(days=n_days)
+        maintenances = self.queryset
+        return_list = []
+        for maintenance in maintenances:
+            if end_date <= maintenance.start_time.date() < today:
+                return_list.append(maintenance)
+
+        return return_list
+
+    def _get_historical_matrix(self):
+        """Gets the historical matrix of the past maintenances.
+
+        Returns:
+            dict: A dictionary that represents the historical matrix maintenance record data.
+
+            {
+                "7 Days": count of past 7 days of maintenance,
+                "30 Days": count of past 30 days of maintenance,
+                "365 Days": count of past 30 days of maintenances
+            }
+        """
+        # TODO: Move to a generic function set up, since this is something that should be exposed via the Capacity
+        #       Metrics plugin when enabled.
+        return_dict = {
+            "past_7_days_maintenance": self.get_maintenance_past_n_days(-7),
+            "past_30_days_maintenance": self.get_maintenance_past_n_days(-30),
+            "past_365_days_maintenance": self.get_maintenance_past_n_days(-365),
+        }
+        return return_dict
+
+    def calculate_future_maintenances(self):
+        """Method to calculate future maintenances.
+
+        Returns:
+            int: Count of future maintenances
+        """
+        today = datetime.date.today()
+        count = 0
+        for ckt_maint in self.queryset:
+            if ckt_maint.start_time.date() > today:
+                count += 1
+
+        return count
+
+    @staticmethod
+    def get_month_list():
```

Comment:
adding some basic unittesting for all these functions would be great

Reply:
I agree. I need to put some brain power into this one again. It has just been a bit too long for me at the moment. It will probably be better than the manual testing that I have been doing then too.

Reply:
I need to add a few more tests tomorrow, including the next couple that we see below. I'm excited to do some more of these.